### PR TITLE
vtsls: init at 0.2.6

### PR DESCRIPTION
**Pull request by kuglimon**
Provides [vtsls](https://github.com/yioneko/vtsls), a typescript language server.

I've lightly used this for a couple of weeks on NixOS and nix-darwin.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by kuglimon** [comment]
`nix-build --check` complains and it's due to some lock file pnpm creates, not sure if those can be solved. It'll add fields `prunedAt` and `storeDir` to `node_modules/.modules.yaml` and those change from build to build.

File: pkgs/by-name/vt/vtsls/package.nix [hunk 1]
```diff
@@ -0,0 +1,92 @@
+{
+  stdenv,
+  lib,
+  fetchFromGitHub,
+  nodejs_20,
+  git,
+  pnpm_8,
+}:
+
+stdenv.mkDerivation (finalAttrs: rec {
```
Comment (drupol) [inline pkgs/by-name/vt/vtsls/package.nix:10]:
```suggestion
stdenv.mkDerivation (finalAttrs: {
```

File: pkgs/by-name/vt/vtsls/package.nix [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  pname = "vtsls";
+  version = "0.2.6";
+
+  src = fetchFromGitHub {
+    owner = "yioneko";
+    repo = "vtsls";
+    rev = "server-v${version}";
+    sha256 = "sha256-HCi9WLh4IEfhgkQNUVk6IGkQfYagg805Rix78zG6xt0=";
```
Comment (drupol) [inline pkgs/by-name/vt/vtsls/package.nix:18]:
```suggestion
    hash = "sha256-HCi9WLh4IEfhgkQNUVk6IGkQfYagg805Rix78zG6xt0=";
```

File: pkgs/by-name/vt/vtsls/package.nix [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+    runHook postInstall
+  '';
+
+  meta = with lib; {
+    description = "LSP wrapper for typescript extension of vscode.";
+    homepage = "https://github.com/yioneko/vtsls";
+    license = licenses.mit;
+    maintainers = with maintainers; [ kuglimon ];
+    mainProgram = "bin/vtsls";
```
Comment (drupol) [inline pkgs/by-name/vt/vtsls/package.nix:89]:
```suggestion
    mainProgram = "vtsls";
```

**Review by drupol — CHANGES_REQUESTED** [review]

File: pkgs/by-name/vt/vtsls/package.nix (hunk 1 above)
Comment (kuglimon) [reply pkgs/by-name/vt/vtsls/package.nix:10]: Isn't this required for following to work?

```nix
   rev = "server-v${version}";
```

Comment (drupol) [reply pkgs/by-name/vt/vtsls/package.nix:10]: Use `rev = "server-v${finalAttrs.version}";` instead ;)

Comment (kuglimon) [reply pkgs/by-name/vt/vtsls/package.nix]: Ah, sweet. Thanks!

**Comment by kuglimon** [comment]
@drupol rereview ping.

File: pkgs/by-name/vt/vtsls/package.nix [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    owner = "yioneko";
+    repo = "vtsls";
+    rev = "server-v${finalAttrs.version}";
+    hash = "sha256-HCi9WLh4IEfhgkQNUVk6IGkQfYagg805Rix78zG6xt0=";
+    fetchSubmodules = true;
+  };
+
+  nativeBuildInputs = [
+    nodejs_20
+    git
```
Comment (drupol) [inline pkgs/by-name/vt/vtsls/package.nix:24]: Why is git required?
In any case, I think here `gitMinimal` is enough.

File: pkgs/by-name/vt/vtsls/package.nix [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  buildPhase = ''
+    runHook preBuild
+
+    # During build vtsls needs a working git installation.
+    git config --global user.name nixbld
+    git config --global user.email nixbld@example.com
+
+    # This is the sha for submodule packages/service/vscode. Unsure if there's a
+    # better way to get this.
+    echo "65d85f4cd0b4220fe8bb494898e001c7ad20e59b" > ./packages/service/HEAD
```
Comment (drupol) [inline pkgs/by-name/vt/vtsls/package.nix:57]: Where is coming this ID from? This needs to be removed or extracted automatically.

**Review by drupol — CHANGES_REQUESTED** [review]

File: pkgs/by-name/vt/vtsls/package.nix (hunk 4 above)
Comment (kuglimon) [reply pkgs/by-name/vt/vtsls/package.nix:24]: From what I've understood from vtsls build setup is that it mainly uses it to apply patches to the vscode project it pulls as a submodule. It does create some other files with git shas included, but I suspect those are just for debugging.

I could patch out the whole building part and apply those patches using nix, which I did look into. The patches are included in the vtsls repo but their paths are not relative to root. I just didn't know how to apply patches with nix from the repository it's building. And there's the path issue, which is probably solvable by substitution during prePatch hook.

I'll test out `gitMinimal`, thanks!

File: pkgs/by-name/vt/vtsls/package.nix (hunk 5 above)
Comment (kuglimon) [reply pkgs/by-name/vt/vtsls/package.nix:57]: This is the SHA of the vscode submodule. I'll have to check if this is truly needed in our case. A SHA is used during `vtsls` build but I suspect it's just used to skip applying patches every time.

Comment (kuglimon) [reply pkgs/by-name/vt/vtsls/package.nix]: Seems like using a dummy sha works just fine. Requires less patching of the build steps.

**Comment by kuglimon** [comment]
Raised nodejs to 22. I've been running it locally for a week or so without issues. 22 is going to be the next LTS on 29.10, might as well update.

**Comment by kuglimon** [comment]
Rereview ping @drupol

File: pkgs/by-name/vt/vtsls/package.nix [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    runHook preInstall
+
+    mkdir -p $out/{bin,lib/vtsls-language-server}
+    cp -r {packages,node_modules} $out/lib/vtsls-language-server
+    ln -s $out/lib/vtsls-language-server/packages/server/bin/vtsls.js $out/bin/vtsls
+
+    runHook postInstall
+  '';
+
+  meta = with lib; {
```
Comment (drupol) [inline pkgs/by-name/vt/vtsls/package.nix:85]: It would be nice to get rid of this.

```suggestion
  meta = {
```

**Review by drupol — APPROVED** [review]
LGTM overall, but I wish I would have someone else to review it too.

**Review by aqrln — APPROVED** [review]
Built and tested on aarch64-linux, works great

**Comment by KamWithK** [comment]
Used it from the patches and worked great with my NeoVim

**Comment by redxtech** [comment]
Tested and confirmed working for me as well.

**Review by redxtech — APPROVED** [review]
LGTM :+1: 

It would be nice to include the suggestion from drupol, I can re-approve after the change is made if necessary.

**Comment by kuglimon** [comment]
@redtech fixed, thanks!

I'm not that familiar with nix, yet; Should I always avoid `with` expressions? Is it expensive to evaluate or something else?

**Review by drupol — APPROVED** [review]

**Comment by drupol** [comment]
> @redtech fixed, thanks!
> 
> I'm not that familiar with nix, yet; Should I always avoid `with` expressions? Is it expensive to evaluate or something else?

We try to avoid using it in front of attribute sets indeed. I don't have all the reasons at the moment (I'm on the smartphone) but they can be found in the issue queue pretty easily.

**Review by redxtech — APPROVED** [review]

**Pull request merged**